src/testing/util.ts

```diff
@@ -0,0 +1,81 @@
+interface Waiter {
+  done: boolean;
+  wakeUp: () => void;
+}
+
+/**
+ * Implementation of a Semaphore class to control the number of
+ * concurrent operations for a given task.
+ *
+ * https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L352
+ */
+export class Semaphore {
+  private available: number;
+  private waiters: Set<Waiter>;
+
+  constructor(max: number) {
+    this.available = max;
+    this.waiters = new Set<Waiter>();
+  }
+
+  private makeNewWaiter(): { waiter: Waiter; promise: Promise<void> } {
+    const waiter: Waiter = { done: false, wakeUp: () => {} };
+
+    const promise = new Promise<void>((resolve) => {
+      waiter.wakeUp = resolve;
+      this.waiters.add(waiter);
+    }).finally(() => {
+      waiter.done = true;
+    });
+
+    return { waiter, promise };
+  }
+
+  private async acquire(): Promise<void> {
+    if (this.available > 0) {
+      this.available--;
+      return;
+    }
+
+    const { promise, waiter } = this.makeNewWaiter();
+
+    try {
+      try {
+        await promise;
+      } finally {
+        this.waiters.delete(waiter);
+      }
+    } finally {
+      while (this.available > 0) {
+        if (!this.wakeUpNext()) {
+          break;
+        }
+      }
```

Comment:
What is the situation where this is needed?

Wouldn't release always wake up the next waiter?

Reply:
hmm good question, i was following this but yeah i guess i'm not sure why you need this + the logic in release

https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L420-L425

```python
        finally:
            # New waiters may have arrived but had to wait due to FIFO.
            # Wake up as many as are allowed.
            while self._value > 0:
                if not self._wake_up_next():
                    break  # There was no-one to wake up.
```

Reply:
The reason I ask is I was trying to think through if this is going to result in an infinite loop in some situations

Reply:
I guess not if they are using it there